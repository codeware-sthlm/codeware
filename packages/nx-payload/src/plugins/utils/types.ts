export type PayloadPluginOptions = {
  buildTargetName?: string;
  dockerBuildTargetName?: string;
  dockerRunTargetName?: string;
  generateTargetName?: string;
  mongodbTargetName?: string;
  payloadTargetName?: string;
  postgresTargetName?: string;
  serveTargetName?: string;
  startTargetName?: string;
  stopTargetName?: string;
};

export type NormalizedOptions = Required<PayloadPluginOptions>;
