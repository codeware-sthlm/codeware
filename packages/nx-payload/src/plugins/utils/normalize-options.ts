import type { NormalizedOptions, PayloadPluginOptions } from './types';

export const normalizeOptions = (
  options?: PayloadPluginOptions
): NormalizedOptions => ({
  buildTargetName: options?.buildTargetName ?? 'build',
  generateTargetName: options?.generateTargetName ?? 'gen',
  payloadTargetName: options?.payloadTargetName ?? 'payload',
  serveTargetName: options?.serveTargetName ?? 'serve',
  dxDockerBuildTargetName:
    options?.dxDockerBuildTargetName ?? 'dx:docker-build',
  dxDockerRunTargetName: options?.dxDockerRunTargetName ?? 'dx:docker-run',
  dxMongodbTargetName: options?.dxMongodbTargetName ?? 'dx:mongodb',
  dxPostgresTargetName: options?.dxPostgresTargetName ?? 'dx:postgres',
  dxStartTargetName: options?.dxStartTargetName ?? 'dx:start',
  dxStopTargetName: options?.dxStopTargetName ?? 'dx:stop'
});
