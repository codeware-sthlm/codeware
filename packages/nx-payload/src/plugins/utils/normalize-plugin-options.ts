export type PayloadPluginOptions = {
  generateTargetName?: string;
  payloadTargetName?: string;
  payloadGraphqlTargetName?: string;
  dxMongodbTargetName?: string;
  dxPostgresTargetName?: string;
  dxStartTargetName?: string;
  dxStopTargetName?: string;
};

type NormalizedOptions = Required<PayloadPluginOptions>;

/**
 * Apply default values to plugin options that are not set.
 */
export const normalizePluginOptions = (
  options?: PayloadPluginOptions
): NormalizedOptions => ({
  generateTargetName: options?.generateTargetName ?? 'gen',
  payloadTargetName: options?.payloadTargetName ?? 'payload',
  payloadGraphqlTargetName:
    options?.payloadGraphqlTargetName ?? 'payload-graphql',
  dxMongodbTargetName: options?.dxMongodbTargetName ?? 'dx:mongodb',
  dxPostgresTargetName: options?.dxPostgresTargetName ?? 'dx:postgres',
  dxStartTargetName: options?.dxStartTargetName ?? 'dx:start',
  dxStopTargetName: options?.dxStopTargetName ?? 'dx:stop'
});
