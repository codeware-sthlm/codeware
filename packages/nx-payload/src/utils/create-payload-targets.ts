import type { TargetConfiguration } from '@nx/devkit';
import type { RunCommandsOptions } from 'nx/src/executors/run-commands/run-commands.impl';

import type { PayloadTarget } from './definitions';

/**
 * Create Payload targets for a project
 *
 * @param args - arguments
 * @returns The payload targets
 */
export function createPayloadTargets(args: {
  isGraphQLDisabled?: boolean;
  projectName: string;
  projectRoot: string;
}): Record<PayloadTarget, TargetConfiguration> {
  const { isGraphQLDisabled, projectName, projectRoot } = args;

  const dxPrefix = 'Developer experience (DX)';

  return {
    gen: {
      metadata: {
        description:
          'Generate types and GraphQL schema (if enabled) from the Payload configuration',
        technologies: ['node']
      },
      executor: 'nx:run-commands',
      inputs: ['{projectRoot}/src/payload.config.ts'],
      options: {
        commands: isGraphQLDisabled
          ? ['npx payload generate:types']
          : [
              'npx payload generate:types',
              'npx payload-graphql generate:schema'
            ],
        cwd: projectRoot,
        parallel: false
      } satisfies Partial<RunCommandsOptions>
    },
    payload: {
      metadata: {
        description: 'Run a native Payload command',
        technologies: ['node']
      },
      executor: 'nx:run-commands',
      options: {
        command: 'npx payload',
        forwardAllArgs: true,
        cwd: projectRoot
      } satisfies Partial<RunCommandsOptions>
    },
    'payload-graphql': {
      metadata: {
        description: 'Run a native Payload GraphQL command',
        technologies: ['node']
      },
      executor: 'nx:run-commands',
      options: {
        command: 'npx payload-graphql',
        forwardAllArgs: true,
        cwd: projectRoot
      } satisfies Partial<RunCommandsOptions>
    },
    'dx:mongodb': {
      metadata: {
        description: `${dxPrefix} - Start a Mongo database in Docker container`,
        technologies: ['docker']
      },
      command: `docker ps -q -f name=mongodb-${projectName} | grep . && echo '[Running] mongodb is already started' || docker run --name mongodb-${projectName} --rm -d -p 27017:27017 mongo`
    },
    'dx:postgres': {
      metadata: {
        description: `${dxPrefix} - Start a Postgres database in Docker container`,
        technologies: ['docker']
      },
      executor: 'nx:run-commands',
      options: {
        command: `docker ps -q -f name=postgres-${projectName} | grep . && echo '[Running] PostgreSQL init process complete' || docker run --name postgres-${projectName} --rm --env-file ${projectRoot}/.env.local -p 5432:5432 postgres`,
        readyWhen: 'PostgreSQL init process complete'
      } satisfies Partial<RunCommandsOptions>
    },
    'dx:start': {
      metadata: {
        description: `${dxPrefix} - Start Payload admin application and Mongo and Postgres databases in Docker containers`,
        technologies: ['docker']
      },
      command: `docker compose -f ${projectRoot}/docker-compose.yml up -d`
    },
    'dx:stop': {
      metadata: {
        description: `${dxPrefix} - Stop all Docker containers started by 'dx:start' target`,
        technologies: ['docker']
      },
      command: `docker compose -f ${projectRoot}/docker-compose.yml down`
    }
  };
}
