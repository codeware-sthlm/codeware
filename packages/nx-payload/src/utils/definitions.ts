import { type PackageManager, TargetConfiguration } from '@nx/devkit';

/** Map package manager to lock file name */
export const packageManagerLockFiles: Record<PackageManager, string> = {
  bun: 'bun.lockb',
  npm: 'package-lock.json',
  pnpm: 'pnpm-lock.yaml',
  yarn: 'yarn.lock'
};

/**
 * All plugin targets.
 */
export const targets = [
  'build',
  'gen',
  'lint',
  'payload',
  'serve',
  'test',
  'dx:docker-run',
  'dx:docker-build',
  'dx:mongodb',
  'dx:postgres',
  'dx:start',
  'dx:stop'
] as const;

/**
 * All plugin targets.
 *
 * `generate` is aliased to `gen` to avoid conflict with native `nx generate`.
 */
export type Target = (typeof targets)[number];

/**
 * Targets that are generated by the plugin when inference is disabled.
 */
export const generatorTargets = [
  'build',
  'gen',
  'lint',
  'payload',
  'serve',
  'test'
] satisfies Array<Target>;

/**
 * Targets that are generated by the plugin when inference is disabled.
 */
export type GeneratorTarget = (typeof generatorTargets)[number];

/**
 * Targets that can be inferred by the plugin.
 */
export const inferredTargets = [
  'build',
  'gen',
  'payload',
  'serve',
  'dx:docker-run',
  'dx:docker-build',
  'dx:mongodb',
  'dx:postgres',
  'dx:start',
  'dx:stop'
] satisfies Array<Target>;

/**
 * Targets that can be inferred by the plugin.
 *
 * `lint` and `test` are generated by the parent plugin generator.
 */
export type InferredTarget = (typeof inferredTargets)[number];

const dxPrefix = 'Developer experience (DX)';
/**
 * Metadata for inferred targets.
 */
export const metadata: Record<InferredTarget, TargetConfiguration['metadata']> =
  {
    build: {
      description: 'Build the Payload admin application',
      technologies: ['node']
    },
    gen: {
      description:
        'Generate types and GraphQL schema (if enabled) from the Payload configuration',
      technologies: ['node']
    },
    payload: {
      description: 'Run a native Payload command',
      technologies: ['node']
    },
    serve: {
      description: 'Serve Payload admin application in development mode',
      technologies: ['node']
    },
    'dx:docker-run': {
      description: `${dxPrefix} - Run the Docker image from 'dx:docker-build' target`,
      technologies: ['docker']
    },
    'dx:docker-build': {
      description: `${dxPrefix} - Build a Docker image for Payload admin application`,
      technologies: ['docker']
    },
    'dx:mongodb': {
      description: `${dxPrefix} - Start a Mongo database in Docker container`,
      technologies: ['docker']
    },
    'dx:postgres': {
      description: `${dxPrefix} - Start a Postgres database in Docker container`,
      technologies: ['docker']
    },
    'dx:start': {
      description: `${dxPrefix} - Start Payload admin application and Mongo and Postgres databases in Docker containers`,
      technologies: ['docker']
    },
    'dx:stop': {
      description: `${dxPrefix} - Stop all Docker containers started by 'dx:start' target`,
      technologies: ['docker']
    }
  };
