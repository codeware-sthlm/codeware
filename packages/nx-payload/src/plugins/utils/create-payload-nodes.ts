import { dirname, join, relative } from 'path/posix';

import {
  type CreateNodesContext,
  type CreateNodesResult,
  type ProjectConfiguration,
  type TargetConfiguration,
  detectPackageManager,
  normalizePath,
  readJsonFile
} from '@nx/devkit';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { getLockFileName } from '@nx/js';

import { createPayloadTargets } from '../../utils/create-payload-targets';
import { extractPayloadConfig } from '../../utils/extract-payload-config';
import { findUpFs } from '../../utils/find-up-fs';

import {
  type PayloadPluginOptions,
  normalizePluginOptions
} from './normalize-plugin-options';

/**
 * Create Payload nodes for a project
 *
 * @param configFilePath - Path to the Payload config file provided by `createNodesFromFiles`
 * @param context - Create nodes context
 * @param options - Plugin options
 * @param targetsCache - Targets cache
 */
export const createPayloadNodes = async (
  configFilePath: string,
  context: CreateNodesContext,
  options: PayloadPluginOptions | undefined,
  targetsCache: Record<
    string,
    Record<string, TargetConfiguration<PayloadPluginOptions>>
  >
): Promise<CreateNodesResult> => {
  // Convert to POSIX on Windows (workspacePath is os formatted)
  const workspaceRoot = normalizePath(context.workspaceRoot);

  // configFilePath is POSIX formatted
  const configRoot = dirname(configFilePath);
  const fullConfigRoot = join(workspaceRoot, configRoot);

  // Payload config file can be located anywhere, though it's generated initially in the 'src' folder.
  // Find project root by looking for project.json in a parent directory.
  const projectJsonPath = await findUpFs('project.json', {
    startPath: fullConfigRoot,
    stopAtPath: workspaceRoot
  });

  // Do not create targets if project.json isn't found.
  // Plugin limitation: Nx native plugins also allow package.json to be used as a project file.
  if (!projectJsonPath) {
    return {};
  }

  // Create path for project root, e.g. 'apps/my-app'
  const projectRoot = relative(workspaceRoot, dirname(projectJsonPath));

  const { name: projectName } =
    readJsonFile<ProjectConfiguration>(projectJsonPath);

  // Get current GraphQL config state
  const result = extractPayloadConfig(configRoot, 'graphQL.disable');
  if (!result.config) {
    // Don't throw to prevent messing up the targets
    console.error(result.error);
    console.warn(
      `Could not extract Payload config from ${configFilePath}, GraphQL state is not known`
    );
  }
  const isGraphQLDisabled = result?.config?.graphQL?.disable ?? true;

  const hash = await calculateHashForCreateNodes(
    projectRoot,
    normalizePluginOptions(options),
    context,
    [
      getLockFileName(detectPackageManager(workspaceRoot)),
      String(isGraphQLDisabled)
    ]
  );

  // Get payload targets to be inferred for the project
  targetsCache[hash] ??= createPayloadTargets({
    isGraphQLDisabled,
    projectName: String(projectName), // Should have a value by design?
    projectRoot
  });

  return {
    projects: {
      [projectRoot]: {
        root: projectRoot,
        targets: targetsCache[hash]
      }
    }
  };
};
