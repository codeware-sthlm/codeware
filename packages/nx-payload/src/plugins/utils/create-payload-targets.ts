import type {
  CreateNodesContext,
  ProjectConfiguration,
  TargetConfiguration
} from '@nx/devkit';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import type { RunCommandsOptions } from 'nx/src/executors/run-commands/run-commands.impl';

import type { BuildExecutorSchema } from '../../executors/build/schema';
import { metadata } from '../../utils/definitions';

import type { NormalizedOptions } from './types';

export const createPayloadTargets = async (
  projectRoot: string,
  projectConfig: ProjectConfiguration,
  options: NormalizedOptions,
  context: CreateNodesContext
): Promise<Record<string, TargetConfiguration>> => {
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<
    string,
    TargetConfiguration<BuildExecutorSchema | Partial<RunCommandsOptions>>
  > = {};

  // Add build target
  targets[options.buildTargetName] = {
    metadata: metadata.build,
    executor: '@cdwr/nx-payload:build',
    inputs: [
      'default',
      Object.hasOwn(namedInputs, 'production') ? '^production' : '^default'
    ],
    outputs: ['{options.outputPath}'],
    options: {
      main: '{projectRoot}/src/main.ts',
      tsConfig: '{projectRoot}/tsconfig.app.json',
      outputPath: '{workspaceRoot}/dist/{projectRoot}',
      outputFileName: 'src/main.js'
    },
    cache: true
  };

  // Add serve target
  targets[options.serveTargetName] = {
    metadata: metadata.serve,
    executor: '@nx/js:node',
    options: {
      buildTarget: `${projectConfig.name}:${options.buildTargetName}`,
      // Required for file changes to be detected and build target to be re-run
      runBuildTargetDependencies: true,
      watch: true
    }
  };

  // Add generate target
  targets[options.generateTargetName] = {
    metadata: metadata.gen,
    executor: 'nx:run-commands',
    options: {
      commands: [
        'npx payload generate:types',
        'npx payload generate:graphQLSchema'
      ],
      env: {
        PAYLOAD_CONFIG_PATH: '{projectRoot}/src/payload.config.ts'
      },
      parallel: false
    }
  };

  // Add payload target
  targets[options.payloadTargetName] = {
    metadata: metadata.payload,
    executor: 'nx:run-commands',
    options: {
      command: 'npx payload',
      forwardAllArgs: true,
      env: {
        PAYLOAD_CONFIG_PATH: '{projectRoot}/src/payload.config.ts'
      }
    }
  };

  // Add mongodb target
  targets[options.dxMongodbTargetName] = {
    metadata: metadata['dx:mongodb'],
    command: `docker ps -q -f name=mongodb-${projectConfig.name} | grep . && echo '[Running] mongodb is already started' || docker run --name mongodb-${projectConfig.name} --rm -d -p 27017:27017 mongo`
  };

  // Add postgres target
  targets[options.dxPostgresTargetName] = {
    metadata: metadata['dx:postgres'],
    executor: 'nx:run-commands',
    options: {
      command: `docker ps -q -f name=postgres-${projectConfig.name} | grep . && echo '[Running] PostgreSQL init process complete' || docker run --name postgres-${projectConfig.name} --rm --env-file ${projectRoot}/.env.local -p 5432:5432 postgres`,
      readyWhen: 'PostgreSQL init process complete'
    }
  };

  // Add start target
  targets[options.dxStartTargetName] = {
    metadata: metadata['dx:start'],
    command: `docker compose -f ${projectRoot}/docker-compose.yml up -d`
  };

  // Add stop target
  targets[options.dxStopTargetName] = {
    metadata: metadata['dx:stop'],
    command: `docker compose -f ${projectRoot}/docker-compose.yml down`
  };

  // Add docker-build target
  targets[options.dxDockerBuildTargetName] = {
    metadata: metadata['dx:docker-build'],
    command: `docker build -f ${projectRoot}/Dockerfile -t ${projectConfig.name} .`
  };

  // Add docker-run target
  targets[options.dxDockerRunTargetName] = {
    metadata: metadata['dx:docker-run'],
    command: `docker run --name ${projectConfig.name} --rm --env-file ${projectRoot}/.env.local -d -p 3000:3000 ${projectConfig.name}`
  };

  return targets;
};
