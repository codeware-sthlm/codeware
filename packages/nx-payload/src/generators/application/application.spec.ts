import {
  ExpandedPluginConfiguration,
  type Tree,
  readJson,
  readNxJson,
  readProjectConfiguration,
  updateNxJson
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import type { PackageJson } from 'nx/src/utils/package-json';

import { PayloadPluginOptions } from '../../plugins/utils/types';

import generator from './application';
import type { AppGeneratorSchema } from './schema';

describe('application generator', () => {
  let tree: Tree;
  const requiredOptions: AppGeneratorSchema = {
    directory: 'apps/test-dir',
    name: 'test-app',
    skipFormat: true
  };

  console.log = jest.fn();
  console.warn = jest.fn();

  jest.setTimeout(10_000);

  const setInferenceFlag = (useInferencePlugins?: boolean) => {
    const workspace = readNxJson(tree);
    if (useInferencePlugins === undefined) {
      delete workspace.useInferencePlugins;
    } else {
      workspace.useInferencePlugins = useInferencePlugins;
    }
    updateNxJson(tree, workspace);
  };

  const addInferencePlugin = (
    plugin: string | ExpandedPluginConfiguration<PayloadPluginOptions>
  ) => {
    const workspace = readNxJson(tree);
    workspace.plugins = workspace.plugins || [];
    workspace.plugins.push(plugin);
    updateNxJson(tree, workspace);
  };

  beforeEach(() => {
    jest.clearAllMocks();
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it('should add payload dependency', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['payload']).toBeDefined();
  });

  it('should add webpack bundler dependency', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(
      packageJson.dependencies['@payloadcms/bundler-webpack']
    ).toBeDefined();
  });

  it('should add mongodb plugin dependency', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['@payloadcms/db-mongodb']).toBeDefined();
  });

  it('should add postgres plugin dependency', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['@payloadcms/db-postgres']).toBeDefined();
  });

  it('should add richtext slate editor plugin dependency', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(
      packageJson.dependencies['@payloadcms/richtext-slate']
    ).toBeDefined();
  });

  it('should add dependencies for node', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['@nx/node']).toBeUndefined();
    expect(packageJson.devDependencies['@nx/node']).toBeDefined();
  });

  it('should not add dependencies for mongodb or postgres', async () => {
    await generator(tree, requiredOptions);
    const packageJson = readJson<PackageJson>(tree, 'package.json');

    expect(packageJson.dependencies['mongodb']).toBeUndefined();
    expect(packageJson.dependencies['pg']).toBeUndefined();

    expect(packageJson.devDependencies['mongodb']).toBeUndefined();
    expect(packageJson.devDependencies['pg']).toBeUndefined();
  });

  it('should add payload project files', async () => {
    await generator(tree, requiredOptions);

    expect(
      tree.exists(`${requiredOptions.directory}/src/payload.config.ts`)
    ).toBeTruthy();

    expect(
      tree.exists(`${requiredOptions.directory}/eslint.config.mjs`)
    ).toBeTruthy();

    expect(
      tree.exists(`${requiredOptions.directory}/tsconfig.app.json`)
    ).toBeTruthy();
    expect(
      tree.exists(`${requiredOptions.directory}/tsconfig.json`)
    ).toBeTruthy();
    expect(
      tree.exists(`${requiredOptions.directory}/tsconfig.spec.json`)
    ).toBeTruthy();
  });

  it('should exclude folders with run-time generated files on build', async () => {
    await generator(tree, requiredOptions);

    const content: {
      extends: string;
      compilerOptions: object;
      exclude: Array<string>;
      include: Array<string>;
    } = JSON.parse(
      tree.read(`${requiredOptions.directory}/tsconfig.app.json`, 'utf-8')
    );

    expect(
      content.exclude.filter(
        (e) => e === 'src/generated/*.ts' || e === 'src/migrations/*.ts'
      ).length
    ).toBe(2);

    expect(content.extends).toBe('./tsconfig.json');
    expect(Object.keys(content.compilerOptions).length).toBeGreaterThan(0);
    expect(content.include.length).toBeGreaterThan(0);
  });

  it('should add docker files', async () => {
    await generator(tree, requiredOptions);

    expect(
      tree.exists(`${requiredOptions.directory}/docker-compose.yml`)
    ).toBeTruthy();
    expect(tree.exists(`${requiredOptions.directory}/Dockerfile`)).toBeTruthy();
    expect(
      tree.exists(`${requiredOptions.directory}/Dockerfile.dockerignore`)
    ).toBeTruthy();
  });

  it('should create Dockerfile for npm package manager', async () => {
    await generator(tree, requiredOptions);

    const content = tree.read(
      `${requiredOptions.directory}/Dockerfile`,
      'utf-8'
    );
    expect(content).toMatchSnapshot();
  });

  it('should not generate e2e application by default', async () => {
    await generator(tree, requiredOptions);

    expect(
      tree.exists(`${requiredOptions.directory}-e2e/project.json`)
    ).toBeFalsy();
  });

  it('should generate e2e application', async () => {
    await generator(tree, { ...requiredOptions, e2eTestRunner: 'jest' });

    expect(
      tree.exists(`${requiredOptions.directory}-e2e/project.json`)
    ).toBeTruthy();
  });

  it('should add three dotenv files', async () => {
    await generator(tree, requiredOptions);

    const envFiles = tree
      .children(requiredOptions.directory)
      .filter((file) => file.match(/\.env/));

    expect(envFiles).toEqual(['.env.docker', '.env.local', '.env.serve']);
  });

  it('should add folders for auto generated files', async () => {
    await generator(tree, requiredOptions);

    expect(
      tree.exists(`${requiredOptions.directory}/src/generated/.gitkeep`)
    ).toBeTruthy();

    expect(
      tree.exists(`${requiredOptions.directory}/src/migrations/.gitkeep`)
    ).toBeTruthy();
  });

  it('should setup mongodb in payload config by default', async () => {
    await generator(tree, requiredOptions);

    const content = tree.read(
      `${requiredOptions.directory}/src/payload.config.ts`,
      'utf-8'
    );
    expect(content.match(/mongooseAdapter/g).length).toBe(2);
    expect(content.match(/DATABASE_URL/g).length).toBe(1);
  });

  it('should setup mongodb in payload config', async () => {
    await generator(tree, { ...requiredOptions, database: 'mongodb' });

    const content = tree.read(
      `${requiredOptions.directory}/src/payload.config.ts`,
      'utf-8'
    );
    expect(content.match(/mongooseAdapter/g).length).toBe(2);
    expect(content.match(/mongodb:\/\/localhost\/test-app/)).toBeDefined();
    expect(content.match(/postgresAdapter/)).toBeNull();
  });

  it('should setup postgres in payload config', async () => {
    await generator(tree, { ...requiredOptions, database: 'postgres' });

    const content = tree.read(
      `${requiredOptions.directory}/src/payload.config.ts`,
      'utf-8'
    );
    expect(content.match(/postgresAdapter/g).length).toBe(2);
    expect(
      content.match(/postgresql:\/\/postgres:postgres@postgres:5432\/test-app/)
    ).toBeDefined();
    expect(content.match(/mongooseAdapter/)).toBeNull();
  });

  it("should setup plugin inference when 'useInferencePlugins' doesn't exist", async () => {
    setInferenceFlag();
    await generator(tree, requiredOptions);

    const nxJson = readNxJson(tree);
    expect(nxJson.useInferencePlugins).toBeUndefined();
    expect(nxJson.plugins[0]).toMatchObject({
      plugin: '@cdwr/nx-payload/plugin'
    });

    const projectJson = readProjectConfiguration(tree, requiredOptions.name);
    expect(projectJson).toMatchSnapshot();
  });

  it("should setup plugin inference when 'useInferencePlugins' is 'true'", async () => {
    setInferenceFlag(true);
    await generator(tree, requiredOptions);

    const nxJson = readNxJson(tree);
    expect(nxJson.useInferencePlugins).toBe(true);
    expect(nxJson.plugins[0]).toMatchObject({
      plugin: '@cdwr/nx-payload/plugin'
    });

    const projectJson = readProjectConfiguration(tree, requiredOptions.name);
    expect(projectJson).toMatchSnapshot();
  });

  it("should not setup plugin inference when 'useInferencePlugins' is 'false'", async () => {
    setInferenceFlag(false);
    await generator(tree, requiredOptions);

    const nxJson = readNxJson(tree);
    expect(nxJson.useInferencePlugins).toBe(false);
    expect(nxJson.plugins).toBeUndefined();

    const projectJson = readProjectConfiguration(tree, requiredOptions.name);
    expect(projectJson).toMatchSnapshot();
  });

  it('should add plugin defined as object when needed', async () => {
    setInferenceFlag();
    await generator(tree, requiredOptions);

    const nxJson = readNxJson(tree);
    expect(nxJson.plugins).toEqual([
      {
        plugin: '@cdwr/nx-payload/plugin',
        options: {
          buildTargetName: 'build',
          dxDockerBuildTargetName: 'dx:docker-build',
          dxDockerRunTargetName: 'dx:docker-run',
          dxMongodbTargetName: 'dx:mongodb',
          dxPostgresTargetName: 'dx:postgres',
          dxStartTargetName: 'dx:start',
          dxStopTargetName: 'dx:stop',
          generateTargetName: 'gen',
          payloadTargetName: 'payload',
          serveTargetName: 'serve'
        }
      }
    ]);
  });

  it('should skip setup plugin inference when plugin exists as string', async () => {
    addInferencePlugin('@cdwr/nx-payload/plugin');
    await generator(tree, requiredOptions);

    const content = readNxJson(tree);
    expect(content.plugins).toEqual(['@cdwr/nx-payload/plugin']);
  });

  it('should skip setup plugin inference when plugin defined as object without options', async () => {
    addInferencePlugin({ plugin: '@cdwr/nx-payload/plugin' });
    await generator(tree, requiredOptions);

    const content = readNxJson(tree);
    expect(content.plugins).toEqual([{ plugin: '@cdwr/nx-payload/plugin' }]);
  });

  it('should skip setup plugin inference when plugin defined as object with options', async () => {
    addInferencePlugin({
      plugin: '@cdwr/nx-payload/plugin',
      options: {
        buildTargetName: 'my-build',
        generateTargetName: 'my-generate',
        payloadTargetName: 'my-payload',
        serveTargetName: 'my-serve',
        dxDockerBuildTargetName: 'my-docker-build',
        dxDockerRunTargetName: 'my-docker-run',
        dxMongodbTargetName: 'my-mongodb',
        dxPostgresTargetName: 'my-postgres',
        dxStartTargetName: 'my-start',
        dxStopTargetName: 'my-stop'
      }
    });
    await generator(tree, requiredOptions);

    const content = readNxJson(tree);
    expect(content.plugins).toEqual([
      {
        plugin: '@cdwr/nx-payload/plugin',
        options: {
          buildTargetName: 'my-build',
          generateTargetName: 'my-generate',
          payloadTargetName: 'my-payload',
          serveTargetName: 'my-serve',
          dxDockerBuildTargetName: 'my-docker-build',
          dxDockerRunTargetName: 'my-docker-run',
          dxMongodbTargetName: 'my-mongodb',
          dxPostgresTargetName: 'my-postgres',
          dxStartTargetName: 'my-start',
          dxStopTargetName: 'my-stop'
        }
      }
    ]);
  });

  it('should skip setup plugin inference when plugin defined as object with partial options', async () => {
    addInferencePlugin({
      plugin: '@cdwr/nx-payload/plugin',
      options: {
        buildTargetName: 'my-build',
        generateTargetName: 'my-generate',
        payloadTargetName: 'my-payload',
        serveTargetName: 'my-serve'
      }
    });
    await generator(tree, requiredOptions);

    const content = readNxJson(tree);
    expect(content.plugins).toEqual([
      {
        plugin: '@cdwr/nx-payload/plugin',
        options: {
          buildTargetName: 'my-build',
          generateTargetName: 'my-generate',
          payloadTargetName: 'my-payload',
          serveTargetName: 'my-serve'
        }
      }
    ]);
  });

  // @see https://github.com/nrwl/nx/blob/master/packages/remix/src/generators/application/application.impl.spec.ts
  it.todo('should test all options');
});
