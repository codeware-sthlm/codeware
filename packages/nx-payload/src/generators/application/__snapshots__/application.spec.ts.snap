// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application generator should create Dockerfile for npm package manager 1`] = `
"
FROM node:20-alpine as base

FROM base as builder

ARG MONGO_URL
ARG NODE_ENV
ARG PAYLOAD_PUBLIC_SERVER_URL
ARG PORT

ENV MONGO_URL=\${MONGO_URL}
ENV NODE_ENV=\${NODE_ENV}
ENV PAYLOAD_PUBLIC_SERVER_URL=\${PAYLOAD_PUBLIC_SERVER_URL}
ENV PORT=\${PORT}

WORKDIR /app

COPY package.json ./

RUN npm config ls
RUN npm install

COPY . .

# Fix for pnpm workspaces (remove it when not applicable):
# The lock file is hidden from the build process when included in the .gitignore file.
# Since package-lock.json is needed when building the app, we'll remove it from .gitignore.
RUN sed -i '/package-lock.json/d' .gitignore

RUN npx nx build test-app

# We'll create a structure similar to scaffolding a default payload workspace.
# Admin build output should match \`buildPath\` in payload.config.ts.
FROM base as runtime

ENV NODE_ENV production

WORKDIR /app

COPY --from=builder /app/dist/apps/test-dir/package.json ./

RUN npm install --omit=dev

COPY --from=builder /app/dist/apps/test-dir/src   ./dist
COPY --from=builder /app/dist/apps/test-dir/build ./dist/apps/test-dir/build

EXPOSE 3000

CMD ["node", "dist/main.js"]
"
`;

exports[`application generator should not setup plugin inference when 'useInferencePlugins' is 'false' 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "test-app",
  "projectType": "application",
  "root": "apps/test-dir",
  "sourceRoot": "apps/test-dir/src",
  "tags": [],
  "targets": {
    "build": {
      "cache": true,
      "executor": "@cdwr/nx-payload:build",
      "metadata": {
        "description": "Build the Payload admin application",
        "technologies": [
          "node",
        ],
      },
      "options": {
        "main": "apps/test-dir/src/main.ts",
        "outputFileName": "src/main.js",
        "outputPath": "dist/apps/test-dir",
        "tsConfig": "apps/test-dir/tsconfig.app.json",
      },
    },
    "gen": {
      "executor": "nx:run-commands",
      "metadata": {
        "description": "Generate types and GraphQL schema (if enabled) from the Payload configuration",
        "technologies": [
          "node",
        ],
      },
      "options": {
        "commands": [
          "npx payload generate:types",
          "npx payload generate:graphQLSchema",
        ],
        "env": {
          "PAYLOAD_CONFIG_PATH": "{projectRoot}/src/payload.config.ts",
        },
        "parallel": false,
      },
    },
    "lint": {
      "executor": "@nx/eslint:lint",
    },
    "payload": {
      "executor": "nx:run-commands",
      "metadata": {
        "description": "Run a native Payload command",
        "technologies": [
          "node",
        ],
      },
      "options": {
        "command": "npx payload",
        "env": {
          "PAYLOAD_CONFIG_PATH": "{projectRoot}/src/payload.config.ts",
        },
        "forwardAllArgs": true,
      },
    },
    "serve": {
      "executor": "@nx/js:node",
      "metadata": {
        "description": "Serve Payload admin application in development mode",
        "technologies": [
          "node",
        ],
      },
      "options": {
        "buildTarget": "test-app:build",
        "runBuildTargetDependencies": true,
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx/jest:jest",
      "options": {
        "jestConfig": "apps/test-dir/jest.config.ts",
      },
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}",
      ],
    },
  },
}
`;

exports[`application generator should setup plugin inference when 'useInferencePlugins' doesn't exist 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "test-app",
  "projectType": "application",
  "root": "apps/test-dir",
  "sourceRoot": "apps/test-dir/src",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint",
    },
    "test": {
      "executor": "@nx/jest:jest",
      "options": {
        "jestConfig": "apps/test-dir/jest.config.ts",
      },
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}",
      ],
    },
  },
}
`;

exports[`application generator should setup plugin inference when 'useInferencePlugins' is 'true' 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "test-app",
  "projectType": "application",
  "root": "apps/test-dir",
  "sourceRoot": "apps/test-dir/src",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint",
    },
    "test": {
      "executor": "@nx/jest:jest",
      "options": {
        "jestConfig": "apps/test-dir/jest.config.ts",
      },
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}",
      ],
    },
  },
}
`;
